packages/flutter_tools/lib/src/resident_runner.dart

```diff
@@ -1000,6 +985,25 @@ class TerminalHandler {
             await residentRunner.screenshot(device);
         }
         return true;
+      case 'r':
+        if (!residentRunner.canHotReload) {
+          return false;
+        }
+        final OperationResult result = await residentRunner.restart(fullRestart: false);
+        if (!result.isOk) {
+          printStatus('Try again after fixing the above error(s).', emphasis: true);
+        }
+        return true;
+      case 'R':
+        // If hot restart is not supported for all devices, ignore the command.
+        if (!residentRunner.canHotRestart || !residentRunner.hotMode) {
+          return false;
+        }
+        final OperationResult result = await residentRunner.restart(fullRestart: true);
```

Comment:
In analytics there are some unhandled JSON RPC exceptions bubbling up through here. Since you're in this code, would you mind catching those and failing more gracefully?

Reply:
Can you send me link to some of the stack traces?

Reply:
Actually, we're already wrapping the restart operation in a `try/finally` at:

https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/run_hot.dart#L513

It seems like that should extended to include a catch that marks the return as a failure?

Reply:
From digging into the stack trace more, this error is going down a slightly different path in the daemon: 

https://github.com/flutter/flutter/blob/v1.7.8%2Bhotfix.3/packages/flutter_tools/lib/src/commands/daemon.dart#L874

I think we could add error handling to run_hot instead, and pass a `fatal` flag if the error is one of these json rpc exceptions so we can print a better error

Reply:
sgtm